api/src/test/java/org/openmrs/module/queue/api/QueueEntryServiceTest.java

```diff
@@ -109,6 +114,62 @@ public void shouldCreateNewQueueEntryRecord() {
 		assertThat(result.getPriority(), is(conceptPriority));
 	}
 	
+	@Test
+	public void shouldNotCreateDuplicateOverlappingQueueEntryRecords() {
+		Queue queue = mock(Queue.class);
+		QueueEntry queueEntry = mock(QueueEntry.class);
+		
+		Patient patient = mock(Patient.class);
+		Concept conceptStatus = mock(Concept.class);
+		Concept conceptPriority = mock(Concept.class);
+		Date queueStartDate = new Date();
+		
+		when(queueEntry.getQueueEntryId()).thenReturn(QUEUE_ENTRY_ID);
+		when(queueEntry.getQueue()).thenReturn(queue);
+		when(queueEntry.getStatus()).thenReturn(conceptStatus);
+		when(queueEntry.getPriority()).thenReturn(conceptPriority);
+		when(queueEntry.getPatient()).thenReturn(patient);
+		when(queueEntry.getStartedAt()).thenReturn(queueStartDate);
+		when(queueEntry.getEndedAt()).thenReturn(null);
+		when(dao.createOrUpdate(queueEntry)).thenReturn(queueEntry);
+		
+		QueueEntry result = queueEntryService.saveQueueEntry(queueEntry);
+		assertThat(result, notNullValue());
+		assertThat(result.getQueueEntryId(), is(QUEUE_ENTRY_ID));
+		assertThat(result.getQueue(), is(queue));
+		assertThat(result.getStatus(), is(conceptStatus));
+		assertThat(result.getPriority(), is(conceptPriority));
+		assertThat(result.getPatient(), is(patient));
+		assertThat(result.getStartedAt(), is(queueStartDate));
+
+		//attempt to add a second queue entry for the same patient to the same queue
+		Date secondQueueStartDate = new Date();
+		QueueEntry secondQueueEntry = mock(QueueEntry.class);
+		when(secondQueueEntry.getQueueEntryId()).thenReturn(SECOND_QUEUE_ENTRY_ID);
+		when(secondQueueEntry.getQueue()).thenReturn(queue);
+		when(secondQueueEntry.getStatus()).thenReturn(conceptStatus);
+		when(secondQueueEntry.getPriority()).thenReturn(conceptPriority);
+		when(secondQueueEntry.getPatient()).thenReturn(patient);
+		when(secondQueueEntry.getStartedAt()).thenReturn(secondQueueStartDate);
+
+		QueueEntrySearchCriteria searchCriteria = new QueueEntrySearchCriteria();
+		searchCriteria.setPatient(secondQueueEntry.getPatient());
+		searchCriteria.setQueues(Collections.singletonList(queue));
+		//tell the dao to return the first queueEntry that was successfully saved above
+		when(dao.getQueueEntries(searchCriteria)).thenReturn(Collections.singletonList(result));
```

Comment:
This doesn't actually test the most meaningful part of what needs testing - namely that the date overlapping logic works as intended.  Hopefully moving that out to a separate method will facilitate that.